syntax = "proto3";

package user;

message Id {
    string value = 1;
}

message Token {
    string value = 1;
}

message UserSignUpForm {
    string name = 1;
    string email = 2;
    string password = 3;
}

message UserLogInForm {
    string email = 1;
    string password = 2;
}

message UserDTO {
    string name = 1;
    string email = 2;
}

message UserDTOWithToken {
    string name = 1;
    string email = 2;
    string token = 3;
}

message InternalUserDTO {
    string name = 1;
    string email = 2;
    string id = 3;
}

service User {
    /*
    CLIENT ACCESSIBLE Services
    */
    rpc LogIn (UserLogInForm) returns (UserDTOWithToken);
    rpc SignUp (UserSignUpForm) returns (UserDTOWithToken);
    rpc GetMe (Token) returns (UserDTO);


    /*
    MICROSERVICES INTERNAL ONLY Services
    */
    rpc GetId (Token) returns (Id);
    rpc InternalGetUserViaToken (Token) returns (InternalUserDTO);
    rpc InternalGetUserViaId (Id) returns (InternalUserDTO);

    // Admin priviledge services
    // TODO : Implement LATER
    // rpc GetAllUsers ();
    // rpc GetUserById ();
    // rpc RemoveUser ();
}